contracts/dtoken/src/user_flow_protection.rs

```diff
@@ -0,0 +1,125 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::{AccountId, BlockHeight, log, near_bindgen};
+use near_sdk::collections::{UnorderedMap};
+use near_sdk::env::block_height;
+
+const BLOCKS_TO_NEXT_OPERATION: BlockHeight = 100;
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct ActionMutex {
+    blocked_accounts: UnorderedMap<AccountId, BlockHeight>,
+}
+
+impl Default for ActionMutex {
+    fn default() -> Self {
+        Self {
+            blocked_accounts: UnorderedMap::new(b"s".to_vec()),
+        }
+    }
+}
+
+#[near_bindgen]
+impl ActionMutex {
+    pub fn lock(&mut self, account_id: AccountId) {
+        let block_height = block_height();
+        log!("Lock account: {}", account_id);
+        self.blocked_accounts.insert(&account_id, &block_height);
+    }
+
+    pub fn unlock(&mut self, account_id: AccountId) {
+        log!("Unlock operation for account: {}", account_id);
+        self.blocked_accounts.remove(&account_id);
+    }
+
+    pub fn get_last_block_index(&self, account_id: AccountId) -> BlockHeight {
+        self.blocked_accounts.get(&account_id).unwrap_or(0)
+    }
+
+    pub fn try_lock(&mut self, account_id: AccountId) -> bool {
+        log!("Try lock account: {}", account_id);
+        let mut is_locked: bool = true;
+        let current_block_height = block_height();
+        let blocked_index = self.get_last_block_index(account_id.clone());
+        if blocked_index > 0 || current_block_height - blocked_index <= BLOCKS_TO_NEXT_OPERATION {
+            is_locked = false;
```

Comment:
Return required in this case, otherwise 
self.blocked_accounts.insert(&account_id, &current_block_height); 
will be called!